packages/makecode-core/src/commands.ts

```diff
@@ -919,27 +919,62 @@ export async function shareCommand(opts: ProjectOptions) {
 }
 
 export async function getSimHTML(opts: ProjectOptions) {
-    applyGlobalOptions(opts);
-
-    const project = await resolveProject(opts);
-    const cache = await project.getCacheAsync();
-
-    const key = project.editor.website + "-sim.html"
-    return getExpandedPageFromCache(key, cache, project.editor.website);
+    return getExpandedHTML(opts, "cachedSimulatorKey", "sim");
 }
 
 export async function getAssetEditorHTML(opts: ProjectOptions) {
+    return getExpandedHTML(opts, "cachedAssetEditorKey", "asseteditor");
+}
+
+async function getExpandedHTML(opts: ProjectOptions, downloadInfoKey: "cachedSimulatorKey" | "cachedAssetEditorKey", pageName: "sim" | "asseteditor") {
     applyGlobalOptions(opts);
 
     const project = await resolveProject(opts);
     const cache = await project.getCacheAsync();
+    const infoKey = project.editor.website + "-info"
+
+    const rawDownloadInfo = await cache.getAsync(infoKey);
+    const parsed = (rawDownloadInfo ? JSON.parse(host().bufferToString(rawDownloadInfo)) : {}) as downloader.DownloadInfo;
 
-    const key = project.editor.website + "-asseteditor.html"
-    return getExpandedPageFromCache(key, cache, project.editor.website);
+    if (parsed[downloadInfoKey]) {
+        const existing = await cache.getAsync(parsed[downloadInfoKey]);
+        if (existing) {
+            return host().bufferToString(existing);
+        }
+    }
+
+    const key = `${project.editor.website}-${pageName}.html`;
+    const expanded = await getExpandedPageFromCache(key, cache, project.editor.website, project.editor.cdnUrl);
+
+    const expandedKey = `${project.editor.website}-${pageName}Expanded.html`;
+    await cache.setAsync(expandedKey, host().stringToBuffer(expanded));
+
+    // save that this page has been generated. this key is cleared when the cache is updated
+    parsed[downloadInfoKey] = expandedKey;
+    await cache.setAsync(infoKey, host().stringToBuffer(JSON.stringify(parsed)));
+
+    return expanded;
 }
 
-async function getExpandedPageFromCache(cacheKey: string, cache: mkc.Cache, website: string): Promise<string> {
-    const pageHTML = host().bufferToString(await cache.getAsync(cacheKey));
+async function getExpandedPageFromCache(cacheKey: string, cache: mkc.Cache, website: string, cdnUrl: string): Promise<string> {
+    let pageHTML = host().bufferToString(await cache.getAsync(cacheKey));
+
+    const imageUrls: string[] = [];
+    pageHTML = pageHTML.replace(/https:\/\/[\w\/\.\-]+/g, f => {
```

Comment:
Will it ever be the case that the page html would have a url with http rather than https?

Reply:
nope! we should never embed http links in our page; i think that's even a violation of microsoft policy.